src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Services/NuGetProjectManagerService.cs

```diff
@@ -718,5 +719,93 @@ private async ValueTask<T> CatchAndRethrowExceptionAsync<T>(Func<Task<T>> taskFu
                 throw exception;
             }
         }
+
+        public async ValueTask<IReadOnlyDictionary<Tuple<NuGetFramework, string>, IReadOnlyList<IPackageReferenceContextInfo>>> GetTransitivePackageOriginAsync(PackageIdentity transitivePackage, string projectId, CancellationToken ct)
```

Comment:
Would it be easier/cheaper/better to populate this information on one read, through https://github.com/NuGet/NuGet.Client/blob/d2e800b88d6b941732272c3a8cf7115836e98e22/src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Projects/CpsPackageReferenceProject.cs#L215

and something here: https://github.com/NuGet/NuGet.Client/blob/d2e800b88d6b941732272c3a8cf7115836e98e22/src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Projects/ProjectPackages.cs#L15-L16. 


Then you could reuse the existing caching mechanism

Reply:
The reasoning here is that *calculating* the root package for 1 transitive package id is an exhaustive search, so a comparable amount of work would be done. 

Note that this is a suggestion. I'm open to having a longer conversation if necessary.

Reply:
Added targets (from assets file)  and PackageSpec caching for both `GetInstalledAndTransitivePackages()` and `GetTransitiveOrigins()` APIs.

I also added caching `GetTransitiveOrigins()` results, to avoid computing transitive origin twice.

If assets file is changed, cache is invalidated. See `GetFullRestoreGraphAsync()` for context.